classes/logger.php

```diff
@@ -0,0 +1,180 @@
+ * @author      Srdjan Janković <srdjan@catalyst.net.nz>
+ * @copyright   Catalyst IT
+ * @license     http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace tool_heartbeat;
+
+use core_shutdown_manager;
+use moodle_exception, Exception;
+
+defined('MOODLE_INTERNAL') || die();
+
+class logger {
+    private static $classname = __NAMESPACE__.'\logger';
+    protected static $statcallbacks = [];
+
+    /**
+     * Register shutdown handler
+     *
+     * @return void
+     */
+    public static function register_shutdown_handler(): void {
+        core_shutdown_manager::register_function([self::$classname, 'log']);
+    }
+
+    /**
+     * Register log callback
+     *
+     * @param callable $cb
+     * @param ?string  $stream A fopen() compliant string
+     * @return void
+     */
+    public static function register_stat_callback(callable $cb, $stream = null): void {
+        if (!isset(self::$statcallbacks[$stream])) {
+            self::$statcallbacks[$stream] = [];
+        }
+        self::$statcallbacks[$stream][] = $cb;
+    }
+
+    /**
+     * Shutdown handler
+     *
+     * @return void
+     */
+    public static function log(): void {
+        $defaultstream = self::default_log_stream();
+        $statcallbacks = self::$statcallbacks;
+        if ($defaultstream) {
+            if (!isset($statcallbacks[$defaultstream])) {
+                $statcallbacks[$defaultstream] = [];
+            }
+            if (isset($statcallbacks[null])) {
+                $statcallbacks[$defaultstream] = array_merge(
+                    $statcallbacks[$defaultstream],
+                    $statcallbacks[null]
+                );
+            }
+            $statcallbacks[$defaultstream][] = [self::$classname, 'perf_stats'];
+        }
+        unset($statcallbacks[null]);
+
+
+        foreach($statcallbacks as $stream => $callbacks) {
+            $lines = [];
+            foreach ($callbacks as $cb) {
+                try {
+                    $lines = array_merge($lines, call_user_func($cb));
+                } catch (Exception $e) {
+                    error_log($e->getMessage());
+                }
+            }
+            if ($lines) {
+                try {
+                    self::log_to_stream($stream, $lines);
+                } catch (Exception $e) {
+                    error_log($e->getMessage());
+                }
+            }
+        }
+    }
+
+    /**
+     * Log lines to a stream
+     *
+     * @param string $stream A fopen() compliant string
+     * @param array $lines
+     * @return void
+     * @throws moodle_exception
+     */
+    public static function log_to_stream($stream, array $lines): void {
+        global $PAGE;
+
+        if ($fh = fopen($stream, 'a')) {
+            foreach ($lines as $l) {
+                if (fwrite($fh, "{$PAGE->pagetype} - $l") === false) {
+                    throw new moodle_exception("tool_heartbeat\\logger::log_to_stream: Cannot write to $stream");
```

Comment:
I would like to avoid exceptions where we are adding instrumentation to a page and now the page has the possibility of failing due to the instrumentation. I think a debugging() call is a better option here